server/filestore.go

```diff
@@ -2370,7 +2370,41 @@ func (fs *fileStore) Purge() uint64 {
 		cb(-int64(purged), -rbytes, 0, _EMPTY_)
 	}
 
-	return purged
+	return purged, nil
+}
+
+// Compact will remove all messages from this store up to
+// but not including the seq parameter.
+// Will return the number of purged messages.
+func (fs *fileStore) Compact(seq uint64) (uint64, error) {
+	if seq == 0 {
+		return fs.Purge()
+	}
+
+	if fs.isClosed() {
```

Comment:
Why bother? The store could be closed after this call. The functions isClosed(), msgForSeq(), firstSeq() and removeMsg() all individually get rlock/lock, and by the look of it check for closed state (except firstSeq).
To be honest, I think that this function should have locking from start to finish.. which means likely having "locked" versions of those calls.
(what is going to happen during the compact if messages are removed due to limits because new messages are added?)

Reply:
True..

Reply:
I don't agree it would be easier to hold the lock the whole time. Would be all new and custom code.

Reply:
Removed isClosed() check.